rpcs3/Emu/Cell/Modules/cellAudioOut.cpp

```diff
@@ -26,93 +31,260 @@ void fmt_class_string<CellAudioOutError>::format(std::string& out, u64 arg)
 	});
 }
 
-error_code cellAudioOutGetNumberOfDevice(u32 audioOut);
-
-error_code cellAudioOutGetSoundAvailability(u32 audioOut, u32 type, u32 fs, u32 option)
+audio_out_configuration::audio_out_configuration()
 {
-	cellSysutil.warning("cellAudioOutGetSoundAvailability(audioOut=%d, type=%d, fs=0x%x, option=%d)", audioOut, type, fs, option);
+	CellAudioOutSoundMode mode{};
+
+	audio_out& primary_output = out.at(CELL_AUDIO_OUT_PRIMARY);
+	audio_out& secondary_output = out.at(CELL_AUDIO_OUT_SECONDARY);
+
+	std::vector<CellAudioOutSoundMode>& primary_modes = primary_output.sound_modes;
+	std::vector<CellAudioOutSoundMode>& secondary_modes = secondary_output.sound_modes;
+
+	s32 sound_format = (1 << 0); // Linear PCM 2 Ch.
+
+	const psf::registry psf = psf::load_object(fs::file(Emu.GetSfoDir() + "/PARAM.SFO"));
+	if (psf.contains("SOUND_FORMAT")) sound_format = psf.at("SOUND_FORMAT").as_integer();
+
+	const bool supports_lpcm_2   = (sound_format & (1 << 0)); // Linear PCM 2 Ch.
+	const bool supports_lpcm_5_1 = (sound_format & (1 << 2)); // Linear PCM 5.1 Ch.
+	const bool supports_lpcm_7_1 = (sound_format & (1 << 4)); // Linear PCM 7.1 Ch.
+	const bool supports_dts      = (sound_format & (1 << 8)); // DTS 5.1 Ch.
+	const bool supports_ac3      = (sound_format & (1 << 9)); // Dolby Digital 5.1 Ch.
+
+	if (supports_lpcm_2)   cellSysutil.notice("cellAudioOut: found support for Linear PCM 2 Ch.");
+	if (supports_lpcm_5_1) cellSysutil.notice("cellAudioOut: found support for Linear PCM 5.1 Ch.");
+	if (supports_lpcm_7_1) cellSysutil.notice("cellAudioOut: found support for Linear PCM 7.1 Ch.");
+	if (supports_dts)      cellSysutil.notice("cellAudioOut: found support for DTS 5.1 Ch.");
+	if (supports_ac3)      cellSysutil.notice("cellAudioOut: found support for Dolby Digital 5.1 Ch.");
+
+	// TODO: audio_format should be a bitmap, but we'll keep it simple for now (Linear PCM 2 Ch. 48 kHz should always exist)
+	// TODO: more formats:
+	// - Each LPCM with other sample frequencies (we currently only support 48 kHz)
+	// - AAC
+	// - Dolby Digital Plus
+	// - Dolby TrueHD
+	// - DTS-HD High Resolution Audio
+	// - DTS-HD Master Audio
+	// - ...
+	switch (g_cfg.audio.format)
+	{
+	case audio_format::automatic: // Automatic based on supported formats
+	{
+		if (supports_lpcm_2) // Linear PCM 2 Ch.
+		{
+			mode.type = CELL_AUDIO_OUT_CODING_TYPE_LPCM;
```

Comment:
A lot of code repetition here and below, a simple lambda with the 4 params to set would probably be cleaner.

Reply:
better now?

Reply:
:+1: